storage/engine: add rocksDBBatchBuilder

petermattis opened the pull request:
Building batches in Go is significantly faster than performing a cgo call:
~55ns/op vs ~400ns/op. The Go batches do not support indexing, so flush to
a C++ batch whenever a read of the batch is being performed. I experimented
with adding indexing to the Go side, but all attempts were significantly
slower than flushing to C++.

Added an `engine.Batch` interface with a `FreezeFlushes` method which is
called to avoid flushing a batch in order to write the raft applied index
and mvcc stats.

Fixes #6289.
Fixes #6739.

```
name                            old time/op  new time/op  delta
TrackChoices1_Cockroach-32       597µs ± 5%   577µs ± 3%   -3.37%  (p=0.000 n=19+20)
TrackChoices10_Cockroach-32      153µs ± 5%   138µs ± 4%  -10.02%  (p=0.000 n=20+20)
TrackChoices100_Cockroach-32    96.9µs ± 7%  80.5µs ± 5%  -16.95%  (p=0.000 n=20+19)
TrackChoices1000_Cockroach-32   92.0µs ± 7%  73.7µs ± 6%  -19.94%  (p=0.000 n=20+19)
InsertDistinct1_Cockroach-32    2.04ms ± 9%  1.84ms ± 8%   -9.62%  (p=0.000 n=20+20)
InsertDistinct10_Cockroach-32   1.36ms ± 5%  1.23ms ± 4%  -10.09%  (p=0.000 n=19+20)
InsertDistinct100_Cockroach-32  2.03ms ±15%  1.67ms ± 7%  -17.92%  (p=0.000 n=20+19)
KVInsert1_SQL-32                 519µs ± 4%   528µs ± 5%   +1.88%  (p=0.024 n=20+20)
KVInsert10_SQL-32                926µs ± 6%   906µs ± 3%   -2.13%  (p=0.028 n=20+20)
KVInsert100_SQL-32              3.89ms ± 8%  3.53ms ± 8%   -9.14%  (p=0.000 n=20+19)
KVInsert1000_SQL-32             29.4ms ±11%  27.5ms ± 8%   -6.64%  (p=0.000 n=20+20)
KVInsert10000_SQL-32             356ms ±10%   340ms ±11%   -4.41%  (p=0.013 n=19+19)
KVUpdate1_SQL-32                 757µs ± 9%   742µs ± 9%     ~     (p=0.253 n=20+20)
KVUpdate10_SQL-32               1.31ms ± 7%  1.27ms ± 6%   -2.96%  (p=0.001 n=20+19)
KVUpdate100_SQL-32              5.44ms ± 7%  5.24ms ± 7%   -3.74%  (p=0.006 n=20+20)
KVUpdate1000_SQL-32             43.5ms ± 6%  42.0ms ± 9%   -3.62%  (p=0.007 n=20+20)
KVUpdate10000_SQL-32             508ms ±16%   452ms ±14%  -10.97%  (p=0.000 n=20+18)
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6861)

<!-- Reviewable:end -->

petermattis committed:
storage/engine: add rocksDBBatchBuilder

Building batches in Go is significantly faster than performing a cgo
call: ~55ns/op vs ~400ns/op. The Go batches do not support indexing, so
flush to a C++ batch whenever a read of the batch is being performed. I
experimented with adding indexing to the Go side, but all attempts were
significantly slower than flushing to C++.

Added an `engine.Batch` interface with a `Distinct` method which returns
a view of the existing batch which passes reads directly to the
underlying engine. This avoids flushing buffered mutations to C++ such
as the writes for the raft applied index and mvcc stats which occur at
the end of every request.

Fixes #6289.
Fixes #6739.

name                            old time/op  new time/op  delta
TrackChoices1_Cockroach-32       575µs ±16%   562µs ± 3%   -2.14%  (p=0.003 n=16+20)
TrackChoices10_Cockroach-32      132µs ± 2%   125µs ± 3%   -5.03%  (p=0.000 n=20+20)
TrackChoices100_Cockroach-32    78.7µs ± 3%  69.4µs ± 2%  -11.84%  (p=0.000 n=20+19)
TrackChoices1000_Cockroach-32   77.1µs ± 5%  62.5µs ± 3%  -19.01%  (p=0.000 n=20+19)
InsertDistinct1_Cockroach-32    1.73ms ± 9%  1.63ms ±10%   -5.67%  (p=0.002 n=20+20)
InsertDistinct10_Cockroach-32   1.25ms ± 2%  1.18ms ± 2%   -5.59%  (p=0.000 n=19+20)
InsertDistinct100_Cockroach-32  1.70ms ± 2%  1.63ms ± 2%   -4.11%  (p=0.000 n=18+17)
KVInsert1_SQL-32                 513µs ± 2%   512µs ± 2%     ~     (p=0.883 n=20+20)
KVInsert10_SQL-32                865µs ± 1%   852µs ± 2%   -1.55%  (p=0.000 n=18+20)
KVInsert100_SQL-32              3.12ms ± 4%  3.05ms ± 4%   -2.03%  (p=0.001 n=19+19)
KVInsert1000_SQL-32             23.8ms ± 9%  22.7ms ± 9%   -4.88%  (p=0.001 n=20+20)
KVInsert10000_SQL-32             255ms ± 9%   235ms ±10%   -7.84%  (p=0.000 n=18+17)

petermattis commented:
Review status: 0 of 14 files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/replica.go, line 1856 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIUIdDQOzTidl34s7_4:-KIUIdDQOzTidl34s7_5:-1787567903) ([raw file](https://github.com/cockroachdb/cockroach/blob/7faa9f846dba58883c29ebb04f83e3366c8900b8/storage/replica.go#L1856)):*
> ```Go
> 		// distinct spans so we don't need the batch to read its own writes.
> 		if b, ok := batch.(engine.Batch); ok {
> 			b.FreezeFlushes()
> ```

I'm not terribly happy with either of the uses of `FreezeFlushes`, but without them we lose the performance benefit here. I'm open to suggestions for how to avoid them or make them safer or some other alternative.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/replica.go, line 1666 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIW0cTkJZf_EiXUz8_4:-KIW0cTkJZf_EiXUz8_5:1310269292) ([raw file](https://github.com/cockroachdb/cockroach/blob/7faa9f846dba58883c29ebb04f83e3366c8900b8/storage/replica.go#L1666)):*
> ```Go
> 
> 	// Freeze batch flushes as the only remaining use of the batch is for
> 	// range-local keys which we know have not been previously written.
> ```

The applied index key has definitely been written; it's overwritten every time. Where are the reads that you're avoiding with this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/replica.go, line 1666 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIW0cTkJZf_EiXUz8_4:-KIXiP3SBMoEQGbU_TKj:-215162612) ([raw file](https://github.com/cockroachdb/cockroach/blob/7faa9f846dba58883c29ebb04f83e3366c8900b8/storage/replica.go#L1666)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The applied index key has definitely been written; it's overwritten every time. Where are the reads that you're avoiding with this?
</details>
The comment might be confusing. I'm trying to avoid a read of data written to the batch which requires flushing the batch from Go down into C++. We're definitely still reading previously written state, just not state that was written within this batch. As I mention below, this API is annoying. It would be better if I could figure out how to make this performant without doing so, but so far I haven't been able to.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/replica.go, line 1666 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIW0cTkJZf_EiXUz8_4:-KIXy_KSElJ7Phep5NLl:-317255044) ([raw file](https://github.com/cockroachdb/cockroach/blob/7faa9f846dba58883c29ebb04f83e3366c8900b8/storage/replica.go#L1666)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The comment might be confusing. I'm trying to avoid a read of data written to the batch which requires flushing the batch from Go down into C++. We're definitely still reading previously written state, just not state that was written within this batch. As I mention below, this API is annoying. It would be better if I could figure out how to make this performant without doing so, but so far I haven't been able to.
</details>
Ah, right. The comment should say "not been previously written within this batch", then. But still, where exactly are any reads occurring?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/replica.go, line 1666 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIW0cTkJZf_EiXUz8_4:-KIXzE_1ewXTkRCozFRZ:1727383423) ([raw file](https://github.com/cockroachdb/cockroach/blob/7faa9f846dba58883c29ebb04f83e3366c8900b8/storage/replica.go#L1666)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Ah, right. The comment should say "not been previously written within this batch", then. But still, where exactly are any reads occurring?
</details>
Every `MVCCPut` does a read of the previous data (unless we're doing a "blind" put). For non-versioned keys like the applied index this read is necessary to keep the MVCC stats up to date.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Adding a `FreezeFlushes` that lasts for the remainder of the batch seems risky. What if instead we added a variant of `MVCCPut` that takes two `Engines`, one for reads and one for writes? They'd normally be the same, but if we know the value cannot be found in the batch it would pass the underlying engine instead. This would take some more plumbing, but it would avoid the possibility of modifying the behavior of an unrelated operation. 

<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> #### storage/engine: add rocksDBBatchBuilder
> 
> Building batches in Go is significantly faster than performing a cgo call:
> ~55ns/op vs ~400ns/op. The Go batches do not support indexing, so flush to
> a C++ batch whenever a read of the batch is being performed. I experimented
> with adding indexing to the Go side, but all attempts were significantly
> slower than flushing to C++.
> 
> Added an `engine.Batch` interface with a `FreezeFlushes` method which is
> called to avoid flushing a batch in order to write the raft applied index
> and mvcc stats.
> 
> Fixes #6289.
> Fixes #6739.
> 
> ```
> name                            old time/op  new time/op  delta
> TrackChoices1_Cockroach-32       597µs ± 5%   577µs ± 3%   -3.37%  (p=0.000 n=19+20)
> TrackChoices10_Cockroach-32      153µs ± 5%   138µs ± 4%  -10.02%  (p=0.000 n=20+20)
> TrackChoices100_Cockroach-32    96.9µs ± 7%  80.5µs ± 5%  -16.95%  (p=0.000 n=20+19)
> TrackChoices1000_Cockroach-32   92.0µs ± 7%  73.7µs ± 6%  -19.94%  (p=0.000 n=20+19)
> InsertDistinct1_Cockroach-32    2.04ms ± 9%  1.84ms ± 8%   -9.62%  (p=0.000 n=20+20)
> InsertDistinct10_Cockroach-32   1.36ms ± 5%  1.23ms ± 4%  -10.09%  (p=0.000 n=19+20)
> InsertDistinct100_Cockroach-32  2.03ms ±15%  1.67ms ± 7%  -17.92%  (p=0.000 n=20+19)
> KVInsert1_SQL-32                 519µs ± 4%   528µs ± 5%   +1.88%  (p=0.024 n=20+20)
> KVInsert10_SQL-32                926µs ± 6%   906µs ± 3%   -2.13%  (p=0.028 n=20+20)
> KVInsert100_SQL-32              3.89ms ± 8%  3.53ms ± 8%   -9.14%  (p=0.000 n=20+19)
> KVInsert1000_SQL-32             29.4ms ±11%  27.5ms ± 8%   -6.64%  (p=0.000 n=20+20)
> KVInsert10000_SQL-32             356ms ±10%   340ms ±11%   -4.41%  (p=0.013 n=19+19)
> KVUpdate1_SQL-32                 757µs ± 9%   742µs ± 9%     ~     (p=0.253 n=20+20)
> KVUpdate10_SQL-32               1.31ms ± 7%  1.27ms ± 6%   -2.96%  (p=0.001 n=20+19)
> KVUpdate100_SQL-32              5.44ms ± 7%  5.24ms ± 7%   -3.74%  (p=0.006 n=20+20)
> KVUpdate1000_SQL-32             43.5ms ± 6%  42.0ms ± 9%   -3.62%  (p=0.007 n=20+20)
> KVUpdate10000_SQL-32             508ms ±16%   452ms ±14%  -10.97%  (p=0.000 n=20+18)
> ```
> 
> 
> 
</details>


---

Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-:-KIZthsKLokSNqHtNd49:1850486356)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Ok, so passing two engines to `MVCCPut` would effectively limit the freezing to put operations. Let me see what is involved in doing that.

<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Adding a `FreezeFlushes` that lasts for the remainder of the batch seems risky. What if instead we added a variant of `MVCCPut` that takes two `Engines`, one for reads and one for writes? They'd normally be the same, but if we know the value cannot be found in the batch it would pass the underlying engine instead. This would take some more plumbing, but it would avoid the possibility of modifying the behavior of an unrelated operation. 
</details>


---

Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-:-KIby1tBSt4lUQWv_EU4:-1878652305)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Passing another parameter to (or making another variant of) `MVCC{,Conditional}Put` turned out to be a pain. Instead, I took another approach which creates a new batch type which passes reads to the underlying engine instead of reading from the batch. So `FreezeFlushes` is gone and replaced with this new "distinct" batch. I'm not terribly fond of the terminology, but wanted to see if this approach was more palatable before polishing it more.

PS It looks like there is a small performance gain to be had in `Replica.handleRaftReady` by using the "distinct" batch. Is it safe to use in that context?

<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Ok, so passing two engines to `MVCCPut` would effectively limit the freezing to put operations. Let me see what is involved in doing that.
</details>


---

Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-:-KIeTBsCGl40TXJu_I2U:1851221975)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Looks good. Yes, this should be safe to use in `handleRaftReady`.

<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Passing another parameter to (or making another variant of) `MVCC{,Conditional}Put` turned out to be a pain. Instead, I took another approach which creates a new batch type which passes reads to the underlying engine instead of reading from the batch. So `FreezeFlushes` is gone and replaced with this new "distinct" batch. I'm not terribly fond of the terminology, but wanted to see if this approach was more palatable before polishing it more.
> 
> PS It looks like there is a small performance gain to be had in `Replica.handleRaftReady` by using the "distinct" batch. Is it safe to use in that context?
</details>


---

Reviewed 2 of 4 files at r6.
Review status: 2 of 9 files reviewed at latest revision, 11 unresolved discussions.

---

*[storage/replica.go, line 1669 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIhskJF4u17y1pfxXNS:-KIhskJGdzpZlGidTMFI:739145390) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/replica.go#L1669)):*
> ```Go
> 	// remaining writes are the raft applied index and the updated MVCC stats.
> 	if b, ok := batch.(engine.Batch); ok {
> 		batch = b.Distinct()
> ```

I was thinking that we'd pass `b.Distinct()` to `setAppliedIndex` and `MergeMVCCStats`, but leave the local variable as the original batch (and `Commit` that batch instead of the `Distinct` one). The point of removing `FreezeFlushes` was to avoid making an irreversible shift into `Distinct` mode.

---

*[storage/replica.go, line 1844 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIhtCm16DgW3Z5ppt2U:-KIhtCm16DgW3Z5ppt2V:2028069811) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/replica.go#L1844)):*
> ```Go
> 	// will require restart or retry, execute as normal.
> 	if r.store.TestingKnobs().DisableOnePhaseCommits || !isOnePhaseCommit(ba) {
> 		ba.Header.DistinctSpans = false
> ```

Why do we clear `DistinctSpans` here?  It's already set conservatively so it will be false for non-one-phase transactions. 

---

*[storage/replica.go, line 1886 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIhtR2nHhO9wPPftAR4:-KIhtR2nHhO9wPPftAR5:-499318459) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/replica.go#L1886)):*
> ```Go
> 	batch.Close()
> 	batch = r.store.Engine().NewBatch()
> 	ba.Header.DistinctSpans = false
> ```

Ditto

---

*[storage/engine/batch.go, line 44 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIhutzkjZK1it4QdNWn:-KIhutzkjZK1it4QdNWo:758650649) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/batch.go#L44)):*
> ```Go
> }
> 
> func (b *rocksDBBatchBuilder) Finish() []byte {
> ```

Comment that the builder may be reused after `Finish()`, but that the returned slice is invalidated by any reuse.

---

*[storage/engine/batch.go, line 83 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIhvfesmK3vmXXJQmLY:-KIhvfesmK3vmXXJQmLZ:1130063236) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/batch.go#L83)):*
> ```Go
> 
> func putUint32(b []byte, v uint32) {
> 	b[0] = byte(v >> 24)
> ```

Is it really little-endian in the header and big-endian here? Ew.

---

*[storage/engine/batch.go, line 100 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIixzRUOW1ZPUHc9oBE:-KIixzRUOW1ZPUHc9oBF:-338890139) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/batch.go#L100)):*
> ```Go
> }
> 
> func (b *rocksDBBatchBuilder) encodeKey(key MVCCKey, extra int) {
> ```

Do we not have an existing key encoding function we can use here (I guess the one we have is only in c++)? If not, we need a test that this encoding is in fact the same as the encoding used in c++

---

*[storage/engine/rocksdb.go, line 518 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIj1E82l9R186rJNWid:-KIj1E83C2swcjnFBlhk:-94169492) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L518)):*
> ```Go
> }
> 
> func (r *distinctIterator) Seek(key MVCCKey) {
> ```

Can you remove these forwarding methods, to make it clear that the only thing that's changing is the override of `Close`?

There's nothing especially "distinct" about this; maybe call it `reusableIterator` and replace `batch` with a boolean `inuse`? That batch pointer raises questions in my mind about how it might be use, but it appears to only be tested for nil vs non-nil.

---

*[storage/engine/rocksdb.go, line 678 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIj4UMAumfFo094smu9:-KIj4UMAumfFo094smuA:800022913) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L678)):*
> ```Go
> }
> 
> // ApplyBatchRepr is illegal for a batch and returns an error.
> ```

Update this comment. 

---

*[storage/engine/rocksdb.go, line 745 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIj5ie72HSRrsFqW-m2:-KIj5ie72HSRrsFqW-m3:1057485676) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L745)):*
> ```Go
> 	}
> 
> 	if r.flushes > 0 {
> ```

This block could use some comments; on first reading I thought the `else` was incorrect because there could be both flushed and unflushed changes. 

---

*[storage/engine/rocksdb.go, line 783 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIj6BPtKlOMpb0rLJi3:-KIj6BPtKlOMpb0rLJi4:898037647) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L783)):*
> ```Go
> 
> func (r *rocksDBBatch) Distinct() Engine {
> 	r.distinct.rocksDBBatch = r
> ```

Could this be set in `newRocksDBBatch` instead of overwriting it every time?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-:-KIj6cjqQOA4sGVDYSJA:-2138926779)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 8 files at r4, 1 of 5 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[storage/engine/engine.go, line 185 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIjSFxMeDjGlC5YgvKP:-KIjSFxN8J28W2sOnI-y:-1486888419) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/engine.go#L185)):*
> ```Go
> type Batch interface {
> 	Engine
> 	Distinct() Engine
> ```

Needs commentary.

---

*[storage/engine/rocksdb.go, line 551 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIjSii12c3rmDWvwj8q:-KIjSii12c3rmDWvwj8r:1945949870) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L551)):*
> ```Go
> // that the returned iterator is cached and re-used for the lifetime of the
> // batch. A panic will be thrown if multiple prefix or normal (non-prefix)
> // iterators are used simultaneously on the same batch.
> ```

I don't understand the comment. I can only obtain an iterator through `NewIterator`, right? And it looks like I can have one prefix and one normal iterator at the same time and both are cached. How would I get a second one?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 7 of 7 files at r1, 9 of 9 files at r3, 2 of 8 files at r4, 4 of 5 files at r5.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[storage/engine/rocksdb.go, line 510 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KImNjQyBTZ4S8hKFfl_:-KImNjQyBTZ4S8hKFfla:2147151118) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L510)):*
> ```Go
> 
> func (r *distinctIterator) Close() {
> 	// distinctIterator.Close() is a no-op. The iterator is left open until
> ```

Doesn't look like a no-op to me.

---

*[storage/engine/rocksdb.go, line 551 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIjSii12c3rmDWvwj8q:-KImOklFRJEWQIgmPKEk:1719056762) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L551)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I don't understand the comment. I can only obtain an iterator through `NewIterator`, right? And it looks like I can have one prefix and one normal iterator at the same time and both are cached. How would I get a second one?
</details>
ah, I see. `NewIterator(false); NewIterator(false);` would panic. Are you disallowing that because you're worried that the two iterators are then used concurrently?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I'll make the `handleRaftReady` change in a future PR.

<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Looks good. Yes, this should be safe to use in `handleRaftReady`.
</details>


---

Review status: 4 of 9 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1669 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIhskJF4u17y1pfxXNS:-KInMpo9nP_-RtYfCMW6:-1897675892) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/replica.go#L1669)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I was thinking that we'd pass `b.Distinct()` to `setAppliedIndex` and `MergeMVCCStats`, but leave the local variable as the original batch (and `Commit` that batch instead of the `Distinct` one). The point of removing `FreezeFlushes` was to avoid making an irreversible shift into `Distinct` mode.
</details>
That's cleaner. Done.

---

*[storage/replica.go, line 1844 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIhtCm16DgW3Z5ppt2U:-KInMsUGJt7vU_cA09QN:2036230642) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/replica.go#L1844)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why do we clear `DistinctSpans` here?  It's already set conservatively so it will be false for non-one-phase transactions. 
</details>
I'm not sure what I was thinking. Removed.

---

*[storage/replica.go, line 1886 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIhtR2nHhO9wPPftAR4:-KInN0_2oetOcmSfdfYu:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/replica.go#L1886)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Ditto
</details>
Done.

---

*[storage/engine/batch.go, line 44 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIhutzkjZK1it4QdNWn:-KInNIF6OYQBdSWxKYV4:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/batch.go#L44)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Comment that the builder may be reused after `Finish()`, but that the returned slice is invalidated by any reuse.
</details>
Done.

---

*[storage/engine/batch.go, line 83 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIhvfesmK3vmXXJQmLY:-KInN_MyPwibqajH8tPc:-1235300679) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/batch.go#L83)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Is it really little-endian in the header and big-endian here? Ew.
</details>
The header is what rocksdb expects. The `putUint{32,64}` methods are what we use for writing the walltime and logical portion of the timestamp. Yes, we're using a different endianness than rocksdb. 

---

*[storage/engine/batch.go, line 100 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIixzRUOW1ZPUHc9oBE:-KInNpF4aNbr3wwWCtIx:2090884407) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/batch.go#L100)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Do we not have an existing key encoding function we can use here (I guess the one we have is only in c++)? If not, we need a test that this encoding is in fact the same as the encoding used in c++
</details>
No, we don't have an existing function. `TestBatchBuilder` tests both the key encoding as well as the batch construction. Are you looking for something more specific?

---

*[storage/engine/engine.go, line 185 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIjSFxMeDjGlC5YgvKP:-KInOaMCNRqwAziloCNC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/engine.go#L185)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Needs commentary.
</details>
Done.

---

*[storage/engine/rocksdb.go, line 510 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KImNjQyBTZ4S8hKFfl_:-KInOdcKUDnOtWvx32iF:1809109725) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L510)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Doesn't look like a no-op to me.
</details>
It's a no-op in the sense that we don't close the underlying rocksdb iterator. Let me adjust the comment.

---

*[storage/engine/rocksdb.go, line 518 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIj1E82l9R186rJNWid:-KInOlLXSHVB54IAwzrt:-316014630) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L518)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Can you remove these forwarding methods, to make it clear that the only thing that's changing is the override of `Close`?
> 
> There's nothing especially "distinct" about this; maybe call it `reusableIterator` and replace `batch` with a boolean `inuse`? That batch pointer raises questions in my mind about how it might be use, but it appears to only be tested for nil vs non-nil.
</details>
Yeah, I don't need these anymore. They were copy-pasta.

Done on the rename and the switch from `batch` to `inuse`.

---

*[storage/engine/rocksdb.go, line 551 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIjSii12c3rmDWvwj8q:-KInPET1YzOYUIulyOTs:752139812) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L551)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> ah, I see. `NewIterator(false); NewIterator(false);` would panic. Are you disallowing that because you're worried that the two iterators are then used concurrently?
</details>
Yes, I'm being conservative here. And the concern isn't about concurrency from different goroutines, but from different parts of the same goroutine that is using a single batch.

---

*[storage/engine/rocksdb.go, line 678 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIj4UMAumfFo094smu9:-KInKqyYiz1jYhFO907_:1764391244) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L678)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Update this comment. 
</details>
Thanks for catching this. Done.

---

*[storage/engine/rocksdb.go, line 745 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIj5ie72HSRrsFqW-m2:-KInM06HsmsAyhEAwgTX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L745)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This block could use some comments; on first reading I thought the `else` was incorrect because there could be both flushed and unflushed changes. 
</details>
Done.

---

*[storage/engine/rocksdb.go, line 783 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIj6BPtKlOMpb0rLJi3:-KInMMal6gA-JewjWLFN:2109517944) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L783)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Could this be set in `newRocksDBBatch` instead of overwriting it every time?
</details>
Yes. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-:-KInPRddzOFKYpJFlJBi:-868205253)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 4 files at r7, 4 of 4 files at r8, 5 of 5 files at r9.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1683 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KInScmhQo-al7WCe-36:-KInScmigOfAtm78s7zG:521497111) ([raw file](https://github.com/cockroachdb/cockroach/blob/27650603ea72d71e1c28c5326ad02239a2b1337b/storage/replica.go#L1683)):*
> ```Go
> 	writer := batch
> 	if b, ok := batch.(engine.Batch); ok {
> 		writer = b
> ```

What's the point of this part of the diff now that you're not using the optimization?

---

*[storage/engine/rocksdb.go, line 551 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIjSii12c3rmDWvwj8q:-KInTI1o0ehKfU8jj1OY:-212577656) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L551)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yes, I'm being conservative here. And the concern isn't about concurrency from different goroutines, but from different parts of the same goroutine that is using a single batch.
</details>
Ok. I asked because it's awkward that a certain internal implementation of an interface has semantics which are pretty specialised, but as long as our usage of the `Distinct()` is well-contained, that's probably the best way to go.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
The final performance numbers are going to look much like r9, right?

<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I'll make the `handleRaftReady` change in a future PR.
</details>


---

Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-:-KInV1eDNLIxtZHbh1UY:1093984838)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1683 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KInScmhQo-al7WCe-36:-KInV92PYxqaCr768pWI:947919912) ([raw file](https://github.com/cockroachdb/cockroach/blob/27650603ea72d71e1c28c5326ad02239a2b1337b/storage/replica.go#L1683)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What's the point of this part of the diff now that you're not using the optimization?
</details>
Oops, I messed that up. I was just about to verify the benchmark numbers which would have caught this. Thanks.

---

*[storage/engine/rocksdb.go, line 551 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIjSii12c3rmDWvwj8q:-KInVOmIr1fLZC7LGEZv:-173677106) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/rocksdb.go#L551)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Ok. I asked because it's awkward that a certain internal implementation of an interface has semantics which are pretty specialised, but as long as our usage of the `Distinct()` is well-contained, that's probably the best way to go.
</details>
Yes, it is awkward, but it seems better to be stricter initially and relax the restriction only if needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
the changes LGTM, but I think testing is spread a little thin given the additional complexity added here (any bugs here would be a nightmare). Some randomized tests for the Go batch writer and some sessions that validate that the write-through behavior in some interesting cases (write on `Distinct()`, read that key from the original batch, ...) would be helpful and not too much work.

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> The final performance numbers are going to look much like r9, right?
</details>


---

Review status: 8 of 9 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[storage/engine/rocksdb.go, line 776 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KInWAP3rdZoP7qEdXgk:-KInWAP3rdZoP7qEdXgl:-1249471752) ([raw file](https://github.com/cockroachdb/cockroach/blob/27650603ea72d71e1c28c5326ad02239a2b1337b/storage/engine/rocksdb.go#L776)):*
> ```Go
> 	if err := r.ApplyBatchRepr(r.builder.Finish()); err != nil {
> 		panic(err)
> 	}
> ```

should

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-:-KInV7rkj6yOSd4ZN4TU:-1750422894)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 8 of 9 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1683 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KInWreh-3cRskuoHhVN:-KInWreiy1YGn-NMgqM6:273948280) ([raw file](https://github.com/cockroachdb/cockroach/blob/7932bbb5e060fa83f9a1a867a8d0685cf136fac4/storage/replica.go#L1683)):*
> ```Go
> 	writer := batch
> 	if b, ok := batch.(engine.Batch); ok {
> 		writer = b.Distinct()
> ```

I hope this really shows during performance because it's making me feel somewhat uncomfortable not having the type system make sure that I'm aware downstream of the very special nature of this batch.

I know this is a late addition, but how do you feel about having `Distinct()` return a different interface (which does not implement `Engine`, for example lacking the `Commit()` method, or some flag method added to Engine/Batch) so that functions that can consume it know that they might be getting a distinct one, hence containing the danger?

A non-distinct batch could masquerade as a distinct one. I suppose that would mean the whole MVCC layer would operate on this abstraction instead, but that seems ok.
I know it's an iffy change, but something like that makes sense to make reasoning about the code less of a nightmare. These optimizations are getting pretty complicated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I'll add some additional testing and ping this PR when it is ready.

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> the changes LGTM, but I think testing is spread a little thin given the additional complexity added here (any bugs here would be a nightmare). Some randomized tests for the Go batch writer and some sessions that validate that the write-through behavior in some interesting cases (write on `Distinct()`, read that key from the original batch, ...) would be helpful and not too much work.
</details>


---

Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1683 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KInWreh-3cRskuoHhVN:-KIo0Rc5ays1As5m-jsv:460687073) ([raw file](https://github.com/cockroachdb/cockroach/blob/7932bbb5e060fa83f9a1a867a8d0685cf136fac4/storage/replica.go#L1683)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I hope this really shows during performance because it's making me feel somewhat uncomfortable not having the type system make sure that I'm aware downstream of the very special nature of this batch.
> 
> I know this is a late addition, but how do you feel about having `Distinct()` return a different interface (which does not implement `Engine`, for example lacking the `Commit()` method, or some flag method added to Engine/Batch) so that functions that can consume it know that they might be getting a distinct one, hence containing the danger?
> 
> A non-distinct batch could masquerade as a distinct one. I suppose that would mean the whole MVCC layer would operate on this abstraction instead, but that seems ok.
> I know it's an iffy change, but something like that makes sense to make reasoning about the code less of a nightmare. These optimizations are getting pretty complicated.
</details>
Yes, it definitely showed up in the performance tests. Mind if I file an issue regarding your suggestion? We could definitely have a paired-down `Engine`-like interface that is passed to the `MVCC*` methods. None of those methods need `Commit` or `Open` or many of the other `Engine` methods. This would also take care of the `TODO` in `storage/engine/engine.go` to split up that interface.

---

*[storage/engine/rocksdb.go, line 776 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KInWAP3rdZoP7qEdXgk:-KIo1-XWhvxk7sdUNSfM:-78918585) ([raw file](https://github.com/cockroachdb/cockroach/blob/27650603ea72d71e1c28c5326ad02239a2b1337b/storage/engine/rocksdb.go#L776)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> should
</details>
should?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-:-KIo11dp2pPcp7eSY4IH:1099809162)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1683 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KInWreh-3cRskuoHhVN:-KIo4GI68mqjJX8C1thC:-981193830) ([raw file](https://github.com/cockroachdb/cockroach/blob/7932bbb5e060fa83f9a1a867a8d0685cf136fac4/storage/replica.go#L1683)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yes, it definitely showed up in the performance tests. Mind if I file an issue regarding your suggestion? We could definitely have a paired-down `Engine`-like interface that is passed to the `MVCC*` methods. None of those methods need `Commit` or `Open` or many of the other `Engine` methods. This would also take care of the `TODO` in `storage/engine/engine.go` to split up that interface.
</details>
In fact, these are the only `Engine` methods used by `MVCC` routines:

```
Clear
Get
Iterate
Merge
NewIterator
Put
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1683 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KInWreh-3cRskuoHhVN:-KIo9tGPNRwaqYhU278a:-451352439) ([raw file](https://github.com/cockroachdb/cockroach/blob/7932bbb5e060fa83f9a1a867a8d0685cf136fac4/storage/replica.go#L1683)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> In fact, these are the only `Engine` methods used by `MVCC` routines:
> 
> ```
> Clear
> Get
> Iterate
> Merge
> NewIterator
> Put
> ```
</details>
Yeah, I'm fine deferring this.

---

*[storage/engine/rocksdb.go, line 776 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KInWAP3rdZoP7qEdXgk:-KIo9nokrjNHsc4dGqVj:1666726490) ([raw file](https://github.com/cockroachdb/cockroach/blob/27650603ea72d71e1c28c5326ad02239a2b1337b/storage/engine/rocksdb.go#L776)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> should?
</details>
... Should not commit unfinished comments. I'm not sure what I wanted to say here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1682 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KJ1GnfeCbMf9iJBsXQG:-KJ1GnfeCbMf9iJBsXQH:-1876648799) ([raw file](https://github.com/cockroachdb/cockroach/blob/616c4b98224eba638a2adeb365922b0cb8184809/storage/replica.go#L1682)):*
> ```Go
> 	// remaining writes are the raft applied index and the updated MVCC stats.
> 	writer := batch
> 	if b, ok := batch.(engine.Batch); ok {
> ```

I'm pretty sure we never get here with any engine besides `engine.Batch` (if we did, we'd break atomicity since we rely on being able to discard a partially-written batch on error), so we could make the methods around here pass Batches instead of Engines and get rid of this check. Don't block the PR for it, but it's something to consider when we clean up these interfaces. 

---

*[storage/engine/batch.go, line 83 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIhvfesmK3vmXXJQmLY:-KJ1H_Qlzw-tVghFNUZE:1202746813) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/batch.go#L83)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The header is what rocksdb expects. The `putUint{32,64}` methods are what we use for writing the walltime and logical portion of the timestamp. Yes, we're using a different endianness than rocksdb. 
</details>
Ah, I missed that the big-endian versions are for MVCCKey encoding, not the rocksdb batch encoding. Comments would be nice: this varint encoding is the one used by rocksdb and is incompatible with the one in util/encoding, while the fixed-length encodings are our own. It's unfortunate that we can't easily share these with util/encoding. 

---

*[storage/engine/batch.go, line 100 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIixzRUOW1ZPUHc9oBE:-KJ1JNDtGrssLKHUpawh:-886060627) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/batch.go#L100)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> No, we don't have an existing function. `TestBatchBuilder` tests both the key encoding as well as the batch construction. Are you looking for something more specific?
</details>
TestBatchBuilder tests that we can construct keys and apply batches, but it doesn't test that we're encoding the data correctly. For example, if you swapped `putUint32` to little-endian, I'm pretty sure TestBatchBuilder wouldn't notice. Something else in the test suite would probably blow up, but that's going to be a difficult failure to diagnose. That test needs to read the data back and make sure it got what it thought it wrote. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1682 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KJ1GnfeCbMf9iJBsXQG:-KJ3AM0kAcZVYK8yqc7d:-1183050200) ([raw file](https://github.com/cockroachdb/cockroach/blob/616c4b98224eba638a2adeb365922b0cb8184809/storage/replica.go#L1682)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I'm pretty sure we never get here with any engine besides `engine.Batch` (if we did, we'd break atomicity since we rely on being able to discard a partially-written batch on error), so we could make the methods around here pass Batches instead of Engines and get rid of this check. Don't block the PR for it, but it's something to consider when we clean up these interfaces. 
</details>
Yep, we always have an `engine.Batch`. I'd like to plumb that through as we clean up these interfaces.

---

*[storage/engine/batch.go, line 83 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIhvfesmK3vmXXJQmLY:-KJ3Byp82dRe9e7oNslA:1008488657) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/batch.go#L83)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Ah, I missed that the big-endian versions are for MVCCKey encoding, not the rocksdb batch encoding. Comments would be nice: this varint encoding is the one used by rocksdb and is incompatible with the one in util/encoding, while the fixed-length encodings are our own. It's unfortunate that we can't easily share these with util/encoding. 
</details>
I'll add a bunch of comments when I add the additional tests @tschottdorf is asking for.

---

*[storage/engine/batch.go, line 100 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIixzRUOW1ZPUHc9oBE:-KJ3BCIUgCaHkZdpaVdm:240733536) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/batch.go#L100)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> TestBatchBuilder tests that we can construct keys and apply batches, but it doesn't test that we're encoding the data correctly. For example, if you swapped `putUint32` to little-endian, I'm pretty sure TestBatchBuilder wouldn't notice. Something else in the test suite would probably blow up, but that's going to be a difficult failure to diagnose. That test needs to read the data back and make sure it got what it thought it wrote. 
</details>
`TestBatchBuilder` builds a batch representation using both the old C++ code and the new Go code and then compares the two representations. I just verified that changing `putUint32` caused the test to fail.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I add `TestBatchDistinct` which caught a buglet in that I wasn't overriding `Engine.Get` for the distinct batch. Now fixed. Added `TestBatchBuilderStress` which randomly generates batches using both the C++ and Go batch building code and compares the representations. PTAL.

<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I'll add some additional testing and ping this PR when it is ready.
</details>


---

Review status: 6 of 9 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-:-KJ5qnP-mR0GR4brvDID:-1464742664)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I add `TestBatchDistinct` which caught a buglet in that I wasn't overriding `Engine.Get` for the distinct batch. Now fixed. Added `TestBatchBuilderStress` which randomly generates batches using both the C++ and Go batch building code and compares the representations. PTAL.
</details>


---

Review status: 6 of 9 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/engine/batch.go, line 100 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KIixzRUOW1ZPUHc9oBE:-KJ6BlKxS9f21Ct8EHCu:839189947) ([raw file](https://github.com/cockroachdb/cockroach/blob/e18aca090f7f686de20cce7c86069786f50c4cef/storage/engine/batch.go#L100)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `TestBatchBuilder` builds a batch representation using both the old C++ code and the new Go code and then compares the two representations. I just verified that changing `putUint32` caused the test to fail.
</details>
Ah, nevermind, I had misread the test and thought it was building one batch from the other. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-:-KJ6D_RKO8CePuQ6XHva:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> :lgtm:
</details>


---

Reviewed 3 of 3 files at r12, 1 of 1 files at r13, 1 of 1 files at r14.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[storage/engine/batch_test.go, line 665 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KJ6JNyMHhiJOw-JfDKo:-KJ6JNyMHhiJOw-JfDKp:-1117868733) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf9cfd5aa15684fac320ae1a317e2577c131ad3c/storage/engine/batch_test.go#L665)):*
> ```Go
> 		t.Fatal(err)
> 	}
> ```

But now if you read from the batch here, you would flush those writes to the RocksDB-batch, in which case `Distinct()` **would** see them, right? That behavior is pretty subtle. Would be good to document it here (and on the interfaces)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-:-KJ6Kb3Yv6x2d5WgjDyl:1417088721)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 7 of 9 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[storage/engine/batch_test.go, line 665 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KJ6JNyMHhiJOw-JfDKo:-KJ6L6WBMa0YRZmX5hzn:607368554) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf9cfd5aa15684fac320ae1a317e2577c131ad3c/storage/engine/batch_test.go#L665)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> But now if you read from the batch here, you would flush those writes to the RocksDB-batch, in which case `Distinct()` **would** see them, right? That behavior is pretty subtle. Would be good to document it here (and on the interfaces)
</details>
That subtlety was not necessary. I've changed `distinctBatch` to always read from the underlying engine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 7 of 9 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[storage/engine/batch_test.go, line 665 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861#-KJ6JNyMHhiJOw-JfDKo:-KJ6N9JnfZtUThJ9bGJK:55364516) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf9cfd5aa15684fac320ae1a317e2577c131ad3c/storage/engine/batch_test.go#L665)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> That subtlety was not necessary. I've changed `distinctBatch` to always read from the underlying engine.
</details>
Ah, I'd missed that. Thanks for clarifying.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 2 files at r15.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6861)*
<!-- Sent from Reviewable.io -->
